File: clients/Scala/src/main/scala/model/Color.scala
package model

import util.StreamUtil

case class Color(r: Float, g: Float, b: Float, a: Float) {
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeFloat(stream, r)
        StreamUtil.writeFloat(stream, g)
        StreamUtil.writeFloat(stream, b)
        StreamUtil.writeFloat(stream, a)
    }
}
object Color {
    def readFrom(stream: java.io.InputStream): Color = Color(
        StreamUtil.readFloat(stream)
        ,
        StreamUtil.readFloat(stream)
        ,
        StreamUtil.readFloat(stream)
        ,
        StreamUtil.readFloat(stream)
        )
}

File: clients/TypeScript/src/model/Color.ts
import {Stream} from './Stream';

export class Color {
    constructor(
        private r: number,
        private g: number,
        private b: number,
        private a: number
    ) {}

    static async readFrom(stream: Stream) {
        const r = await stream.readFloat();
        const g = await stream.readFloat();
        const b = await stream.readFloat();
        const a = await stream.readFloat();

        return new Color(r, g, b, a);
    }
    async writeTo(stream: Stream) {
        await stream.writeFloat(this.r);
        await stream.writeFloat(this.g);
        await stream.writeFloat(this.b);
        await stream.writeFloat(this.a);
    }
}
